test/e2e/xfn_test.go

```diff
@@ -0,0 +1,344 @@
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/envfuncs"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+	"sigs.k8s.io/e2e-framework/third_party/helm"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+	"github.com/crossplane/crossplane/test/e2e/funcs"
+)
+
+func TestXfnRunnerImagePull(t *testing.T) {
+
+	manifests := "test/e2e/manifests/xfnrunner/private-registry/pull"
+	environment.Test(t,
+		features.New("PullFnImageFromPrivateRegistryWithCustomCert").
+			WithLabel(LabelArea, "xfn").
+			WithSetup("InstallRegistryWithCustomTlsCertificate",
+				funcs.AllOf(
+					funcs.AsFeaturesFunc(envfuncs.CreateNamespace("reg")),
+					CreateTLSCertificateAsSecret("private-docker-registry.reg.svc.cluster.local", "reg"),
+					InstallDockerRegistry(),
+				)).
+			WithSetup("CopyFnImageToRegistry", CopyImagesToRegistry()).
+			WithSetup("CrossplaneDeployedWithFunctionsEnabled", CrossplaneDeployedWithFunctionsEnabled()).
+			WithSetup("ProvideNopDeployed", ProvideNopDeployed()).
+			Assess("CompositionWithFunctionIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "composition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "composition.yaml"),
+			)).
+			Assess("ClaimIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+			)).
+			Assess("ClaimBecomesAvailable", funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available())).
+			Assess("ManagedResourcesProcessedByFunction", ManagedResourcedProcessedByFunction()).
+			Feature(),
+	)
+}
+
+func resourceGetter(ctx context.Context, t *testing.T, config *envconf.Config) func(string, string, string, string) *unstructured.Unstructured {
+	return func(name string, namespace string, apiVersion string, kind string) *unstructured.Unstructured {
+		client := config.Client().Resources().GetControllerRuntimeClient()
+		u := &unstructured.Unstructured{}
+		gv, err := schema.ParseGroupVersion(apiVersion)
+		if err != nil {
+			t.Fatal(err)
+		}
+		u.SetGroupVersionKind(gv.WithKind(kind))
+		if err := client.Get(ctx, cr.ObjectKey{Name: name, Namespace: namespace}, u); err != nil {
+			t.Fatal("cannot get claim", err)
+		}
+		return u
+	}
+}
+func resourceValue(t *testing.T, u *unstructured.Unstructured, path ...string) map[string]string {
```

Comment:
Nit: but would be awesome if we could make all these private functions somewhat reusable and move to funcs if we feel it would be worth it, but not a blocker